### ISSUE TITLE
Neural Networks for Bayesian Optimsation?

### ISSUE THREAD

**Issue by MechCoder** [scikit-optimize/scikit-optimize#313]
See: https://arxiv.org/pdf/1502.05700.pdf

**Comment by MechCoder** [comment]
It seems similar to Bayesian Linear Regression except that one replaces the input x with non-linear features of a neural network.. Comparing the last equation in https://www.cs.utah.edu/~fletcher/cs6957/lectures/BayesianLinearRegression.pdf and Eqns 6 and 7 in the paper linked above.

**Comment by betatim** [comment]
<snark>How is 1502.05700 an interesting paper? They just replaced a GP with a different kind of function approximator ... should we write a paper on using each regressor in sklearn as approximator?</snark>

The tricky bit is how to get a uncertainty estimate for the regressor output right? Plugging their NN into a bayesian linear regressor is quite a cute idea to solve that. They postulate that running a full bayesian NN isn't feasible (I think), we should try out the [nice PyMC3 example](https://pymc-devs.github.io/pymc3/notebooks/bayesian_neural_network_advi.html) and [twiecki's blog post](http://twiecki.github.io/blog/2016/06/01/bayesian-deep-learning/) to see what can be done.

Probably worth putting this under "research" as we would have to decide if we really wanted to add PyMC3 as a dependency, etc, etc.

**Comment by MechCoder** [comment]
> Should we write a paper on using each regressor in sklearn as approximator?

If it shows improved performance, then why not?

Do full bayesian Neural Networks give a distribution over all weights in the network? That seems pretty intense lol. I like the idea of using a neural network to model the hyperparameter interactions though before feeding it into a linear bayesian regressor.

Btw according to the paper, our random forest implementation is better than SMAC's (both branin and hart6) and our GP implementation isn't too bad either without slice sampling, the beta kernel etc.

**Comment by glouppe** [comment]
> Do full bayesian Neural Networks give a distribution over all weights in the network? That seems pretty intense lol. 

Yes. Bayesian neural networks define a distribution over neural networks, which amounts in a distribution over their weight parameters. (Much like GPs define a distribution over functions.)

**Comment by glouppe** [comment]
See also https://arxiv.org/pdf/1612.01474.pdf for a simpler derivation of uncertainty from neural nets.
